// app/stores/a11ySettings.js
import {writable} from 'svelte/store';

export const a11ySettings = writable({
	brightness: {
		defaultValue: 100,
		format: 'percentage',
		group: 'color',
		id: 'brightness',
		label: 'Brightness',
		next: 'contrast',
		prev: 'grayscale',
		range: [0, 100],
		value: 100,
	},
	contrast: {
		defaultValue: 100,
		format: 'percentage',
		group: 'color',
		id: 'contrast',
		label: 'Contrast',
		next: null,
		prev: 'brightness',
		range: [0, 100],
		value: 100,
	},
	cvd: {
		currentValueIndex: 0,
		defaultValueIndex: 0,
		group: 'color',
		id: 'cvd',
		label: 'Color Vision Deficiency',
		next: 'hue',
		prev: 'invert',
		value: 'None',
		values: ['None', 'Protanopia', 'Deuteranopia', 'Tritanopia'],
	},
	fontScaling: {
		currentValueIndex: 2,
		defaultValueIndex: 2,
		format: 'percentage',
		group: 'text',
		id: 'fontScaling',
		label: 'Font scale',
		next: 'lineHeight',
		prev: 'typeface',
		value: 100,
		values: [50, 75, 100, 125, 150],
	},
	grayscale: {
		defaultValue: 0,
		format: 'percentage',
		group: 'color',
		id: 'grayscale',
		label: 'Grayscale',
		next: 'brightness',
		prev: 'hue',
		range: [0, 100],
		value: 0,
	},
	hue: {
		defaultValue: 0,
		format: 'degrees',
		group: 'color',
		id: 'hue',
		label: 'Hue',
		next: 'grayscale',
		prev: 'cvd',
		range: [0, 360],
		value: 0,
	},
	invert: {
		defaultValue: false,
		format: 'boolean',
		group: 'color',
		id: 'invert',
		label: 'Invert',
		next: 'cvd',
		prev: 'wordSpacing',
		value: false,
	},
	letterSpacing: {
		currentValueIndex: 0,
		defaultValueIndex: 0,
		format: 'percentage',
		group: 'text',
		id: 'letterSpacing',
		label: 'Letter spacing',
		next: 'wordSpacing',
		prev: 'lineHeight',
		value: 0,
		values: [0, 10, 20],
	},
	lineHeight: {
		currentValueIndex: 1,
		defaultValueIndex: 1,
		format: 'percentage',
		group: 'text',
		id: 'lineHeight',
		label: 'Line height',
		next: 'letterSpacing',
		prev: 'fontScaling',
		value: 125,
		values: [100, 125, 150, 175, 200],
	},
	typeface: {
		currentValueIndex: 0,
		defaultValueIndex: 0,
		group: 'text',
		id: 'typeface',
		label: 'Font',
		next: 'fontScaling',
		prev: null,
		value: 'Archivo',
		values: [
			'Archivo',
			'Avenir Next Variable',
			'Noboto Flex',
			'Open Dyslexia'
		],
	},
	wordSpacing: {
		currentValueIndex: 0,
		defaultValueIndex: 0,
		format: 'percentage',
		group: 'text',
		id: 'wordSpacing',
		label: 'Word spacing',
		next: 'invert',
		prev: 'letterSpacing',
		value: 0,
		values: [0, 20, 40],
	},
});
