// app/stores/a11ySettings.js
import * as _ from 'lamb';
import {derived, get, writable} from 'svelte/store';
import {
	applyFnMap,
	isNot,
	isNotNull,
	makeMergeAppliedFnMap,
	mergeObj,
	makePostfixed
} from '@svizzle/utils';

export const defaultA11ySettings = {
	brightness: {
		defaultValue: 100,
		format: 'percentage',
		group: 'color',
		id: 'brightness',
		label: 'Brightness',
		next: 'contrast',
		prev: 'grayscale',
		range: [0, 100],
		value: 100,
	},
	contrast: {
		defaultValue: 100,
		format: 'percentage',
		group: 'color',
		id: 'contrast',
		label: 'Contrast',
		next: null,
		prev: 'brightness',
		range: [0, 100],
		value: 100,
	},
	cvd: {
		defaultValue: 'None',
		group: 'color',
		id: 'cvd',
		label: 'Color Vision Deficiency',
		next: 'hue',
		prev: 'invert',
		value: 'None',
		values: ['None', 'Protanopia', 'Deuteranopia', 'Tritanopia'],
	},
	fontScaling: {
		defaultValue: 100,
		format: 'percentage',
		group: 'text',
		id: 'fontScaling',
		label: 'Font scale',
		next: 'lineHeight',
		prev: 'typeface',
		value: 100,
		values: [50, 75, 100, 125, 150],
	},
	grayscale: {
		defaultValue: 0,
		format: 'percentage',
		group: 'color',
		id: 'grayscale',
		label: 'Grayscale',
		next: 'brightness',
		prev: 'hue',
		range: [0, 100],
		value: 0,
	},
	hue: {
		defaultValue: 0,
		format: 'degrees',
		group: 'color',
		id: 'hue',
		label: 'Hue',
		next: 'grayscale',
		prev: 'cvd',
		range: [0, 360],
		value: 0,
	},
	invert: {
		defaultValue: false,
		format: 'boolean',
		group: 'color',
		id: 'invert',
		label: 'Invert',
		next: 'cvd',
		prev: 'wordSpacing',
		value: false,
	},
	letterSpacing: {
		defaultValue: 0,
		format: 'percentage',
		group: 'text',
		id: 'letterSpacing',
		label: 'Letter spacing',
		next: 'wordSpacing',
		prev: 'lineHeight',
		value: 0,
		values: [0, 10, 20],
	},
	lineHeight: {
		defaultValue: 150,
		format: 'percentage',
		group: 'text',
		id: 'lineHeight',
		label: 'Line height',
		next: 'letterSpacing',
		prev: 'fontScaling',
		value: 150,
		values: [100, 125, 150, 175, 200],
	},
	typeface: {
		defaultValue: 'Archivo',
		group: 'text',
		id: 'typeface',
		label: 'Font',
		next: 'fontScaling',
		prev: null,
		value: 'Archivo',
		values: [
			'Archivo',
			'Avenir Next Variable',
			'Noboto Flex',
			'Courier New',
			'Open Dyslexia'
		],
	},
	wordSpacing: {
		defaultValue: 0,
		format: 'percentage',
		group: 'text',
		id: 'wordSpacing',
		label: 'Word spacing',
		next: 'invert',
		prev: 'letterSpacing',
		value: 0,
		values: [0, 20, 40],
	},
};

export const a11ySettings = writable(defaultA11ySettings);

/* init */

const isFirstSetting = _.pipe([_.getKey('prev'), _.isNull]);
const getFirstId = _.pipe([
	_.values,
	_.findWhere(isFirstSetting),
	_.getKey('id')
]);
const firstId = getFirstId(defaultA11ySettings);

/* current setting */

export const currentId = writable(firstId);
export const currentSetting = derived(
	[a11ySettings, currentId],
	([settings, id]) => settings[id]
);

/* formatting */

const formats = {
	'percentage': value => `${value}%`,
	'degrees': value => `${value}°`,
	'boolean': value => value ? 'Yes' : 'No',
};
export const formatValue = derived(currentSetting,
	setting => setting.format ? formats[setting.format] : _.identity
);

/* navigation */

export const setNextId = () => currentId.set(get(currentSetting).next);
export const setPrevId = () => currentId.set(get(currentSetting).prev);

export const hasPrev = derived(currentSetting,
	_.pipe([_.getKey('prev'), isNotNull])
);
export const hasNext = derived(currentSetting,
	_.pipe([_.getKey('next'), isNotNull])
);

/* update */

const updateValueOf = (id, value) => {
	a11ySettings.update(_.updateKey(id, mergeObj({value})))
};
export const updateCurrentValue = value => {
	const setting = get(currentSetting);
	updateValueOf(setting.id, value);
};

/* resets */

const getGroupsResetStatus = _.pipe([
	_.values,
	_.groupBy(_.getKey('group')),
	_.mapValuesWith(
		_.pipe([
			_.mapWith(
					_.collect([
						_.getKey('value'),
						_.getKey('defaultValue'),
					])
			),
			_.every(_.apply(_.areSame))
		])
	)
]);
export const groupsResetStatus = derived(a11ySettings, getGroupsResetStatus);

const setValueToDefault = makeMergeAppliedFnMap({value: _.getKey('defaultValue')});
const isNotOfGroup = groupId => _.pipe([_.getKey('group'), isNot(groupId)]);
const resetGroupItems = groupId => _.mapValuesWith(_.adapter([
	_.casus(isNotOfGroup(groupId), _.identity),
	_.casus(_.hasKey('value'), setValueToDefault),
]));
export const resetGroup = groupId => a11ySettings.update(resetGroupItems(groupId));

/* Color corrections CSS property formatter */
const valueFormats = {
	percentage: value => `${value}%`,
	degrees: value => `${value}deg`,
	boolean: value => value ? '100%' : '0%',
}

const includeIfNotDefault = (setting, filter) =>
	setting.defaultValue === setting.value
		? ''
		: `${filter}(${valueFormats[setting.format](setting.value)})`;

const modelFormatters = {
	None: () => '',
	Deuteranopia: () => 'url(#recolor-deuteranopia)',
	Protanopia: () => 'url(#recolor-protanopia)',
	Tritanopia: () => 'url(#recolor-tritanopia)',
};

const getColorCorrection = settings => [
	includeIfNotDefault(settings.invert, 'invert'),
	modelFormatters[settings.cvd.value](settings),
	includeIfNotDefault(settings.hue, 'hue-rotate'),
	includeIfNotDefault(settings.grayscale, 'grayscale'),
	includeIfNotDefault(settings.brightness, 'brightness'),
	includeIfNotDefault(settings.contrast, 'contrast'),
].join(' ');

const getColorStyles = settings =>
	({'--color-correction': getColorCorrection(settings)});

/*
const simulationFormatters = {
	none: () => '',
	deuteranopia: () => 'url(#deuteranopia)',
	protanopia: () => 'url(#protanopia)',
	tritanopia: () => 'url(#tritanopia)',
}
const getCvdSimulation = opts => !opts.active
	? ''
	: simulationFormatters[opts.simulate](opts);

$: $cvdSimulation = ['filter', getCvdSimulation(options)];
*/

/* Text corrections CSS property formatters */

const toRemPercent = _.pipe([_.divideBy(100), makePostfixed('rem')]);

const getTextStyles = applyFnMap({
	'font-family': _.getPath('typeface.value'),
	'font-size': _.pipe([_.getPath('fontScaling.value'), toRemPercent]),
	'letter-spacing': _.pipe([_.getPath('letterSpacing.value'), toRemPercent]),
	'line-height': _.pipe([_.getPath('lineHeight.value'), toRemPercent]),
	'word-spacing': _.pipe([_.getPath('wordSpacing.value'), toRemPercent]),
});

export const a11yTextStyles = derived(a11ySettings, getTextStyles);

export const applyStyles = (domStyle, styles) => {
	_.pairs(styles)
	.forEach(([prop, value]) => {
		domStyle[prop] = value
	});
};
